src/LanguageServer/Protocol/Handler/InlayHint/InlayHintResolveHandler.cs

```diff
@@ -33,39 +34,58 @@ public TextDocumentIdentifier GetTextDocumentIdentifier(LSP.InlayHint request)
         public Task<LSP.InlayHint> HandleRequestAsync(LSP.InlayHint request, RequestContext context, CancellationToken cancellationToken)
         {
             var document = context.GetRequiredDocument();
-            return ResolveInlayHintAsync(document, request, _inlayHintCache, cancellationToken);
+            var options = globalOptions.GetInlineHintsOptions(document.Project.Language);
+            var inlayHintCache = context.GetRequiredService<InlayHintCache>();
+            var resolveData = GetInlayHintResolveData(request);
+            return ResolveInlayHintAsync(document, request, inlayHintCache, resolveData, options, cancellationToken);
         }
 
-        internal static async Task<LSP.InlayHint> ResolveInlayHintAsync(Document document, LSP.InlayHint request, InlayHintCache inlayHintCache, CancellationToken cancellationToken)
+        internal static async Task<LSP.InlayHint> ResolveInlayHintAsync(
+            Document document,
+            LSP.InlayHint request,
+            InlayHintCache inlayHintCache,
+            InlayHintResolveData resolveData,
+            InlineHintsOptions options,
+            CancellationToken cancellationToken)
         {
-            var resolveData = GetInlayHintResolveData(request);
-            var (cacheEntry, inlineHintToResolve) = GetCacheEntry(resolveData, inlayHintCache);
-
             var currentSyntaxVersion = await document.GetSyntaxVersionAsync(cancellationToken).ConfigureAwait(false);
-            var cachedSyntaxVersion = cacheEntry.SyntaxVersion;
+            var resolveSyntaxVersion = resolveData.SyntaxVersion;
 
-            if (currentSyntaxVersion != cachedSyntaxVersion)
+            if (currentSyntaxVersion.ToString() != resolveSyntaxVersion)
             {
-                throw new LocalRpcException($"Cached resolve version {cachedSyntaxVersion} does not match current version {currentSyntaxVersion}")
+                throw new LocalRpcException($"Request resolve version {resolveSyntaxVersion} does not match current version {currentSyntaxVersion}")
                 {
                     ErrorCode = LspErrorCodes.ContentModified
                 };
             }
 
-            var taggedText = await inlineHintToResolve.GetDescriptionAsync(document, cancellationToken).ConfigureAwait(false);
+            var inlineHintToResolve = GetCacheEntry(resolveData, inlayHintCache);
```

Comment:
do we need to audit any other cases?

Reply:
if you're talking about other caches in different handlers, I believe we're good.  We only also use a cache in completion, which doesn't have this same issue - switching editor focus will always dismiss the previous list

Reply:
> switching editor focus will always dismiss the previous list

Perhaps.  Though that seemsl ike a host-specific behavior.  Would we have an issue if the host chnaged and decided to not do that?

Reply:
If the client ever allowed multiple completion sessions at once, then completion will function, but at a slightly degraded experience - the items could be committed, but would not have documentation.  Additionally items with additional text edits (for example unimported types), or complex edits (for example override completion) would only apply the word edit, and not the additional ones.

It may be technically possible to recalculate in completion - but that would slow down the resolve request a fair amount.   IMHO if we want to support the multiple lists at once scenario I'd much prefer adding something to the spec similar to https://github.com/microsoft/language-server-protocol/issues/1802#issuecomment-1805342424 that allows the client to tell us which completion items to hold onto (and for how long).